### PR TITLE
Complete the implementation of type casting

### PR DESCRIPTION
**Pull request by ryzhyk**
- Validate `as` expressions
- Document `as` syntax in the tutorial
- Unary minus implementation
- Tests

### CODE REVIEW

File: doc/tutorial/tutorial.md [hunk 1]
```diff
@@ -558,6 +558,28 @@ numbers respectively:
...
+32'd15 as bit<64>    // type-cast bit<32> to bit<64>
+32'd15 as signed<32> // type-cast bit<32> to signed<32>
+32'd15 as bigint     // type-cast bit<32> to bigint
+```
+
+The following type conversions are currently supported:
+- `bit<N>` to `bit<M>` for any `N` and `M`
+- `signed<N>` to `signed<M>` for any `N` and `M`
+- `bit<N>` to `signed<N>`
+- `signed<N>` to `bit<N>`
```
Comment (mbudiu-vmw) [inline doc/tutorial/tutorial.md]: You will have to explain how overflows are handled. The short explanation is that `signed<N>` uses two's complement representations, and that these casts preserve the bitstring.

File: test/datalog_tests/simple.ast.expected [hunk 2]
```diff
@@ -811,4 +817,19 @@ Cast_u256(.description="264'hfffffffffffffffffffffffffffffffffffffffffffffffffff
...
+Cast_bigint(.description="24'd100  as bigint", .expected=(24'd100 as bigint), .actual=100).
+Cast_bigint(.description="32'd100  as bigint", .expected=(32'd100 as bigint), .actual=100).
+Cast_bigint(.description="16'd100  as bigint", .expected=(16'd100 as bigint), .actual=100).
+Cast_bigint(.description="8'd100   as bigint", .expected=(8'd100 as bigint), .actual=100).
+Cast_bigint(.description="128'd100 as bigint", .expected=(128'd100 as bigint), .actual=100).
+Cast_bigint(.description="256'd100 as bigint", .expected=(256'd100 as bigint), .actual=100).
+Cast_bigint(.description="512'd100 as bigint", .expected=(512'd100 as bigint), .actual=100).
+Cast_bigint(.description="264'hfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffa as bigint", .expected=(264'd29642774844752946028434172162224104410437116074403984394101141506025761187823610 as bigint), .actual=29642774844752946028434172162224104410437116074403984394101141506025761187823610).
+Cast_bigint(.description="(32'sd0 - 32'sd10) as bigint", .expected=((32'sd0 - 32'sd10) as bigint), .actual=(0 - 10)).
+UMinus_s32(.description="-32'sd100", .n=(- 32'sd100)).
```
Comment (mbudiu-vmw) [inline test/datalog_tests/simple.ast.expected]: printing these with a space is a bit confusing

File: doc/tutorial/tutorial.md [hunk 3]
```diff
@@ -558,6 +558,28 @@ numbers respectively:
...
+- `bit<N>` to `bigint`
+- `signed<N>` to `bigint`
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply doc/tutorial/tutorial.md]: There is. We can add it to the standard library.

File: doc/tutorial/tutorial.md (hunk 1 above)
Comment (ryzhyk) [reply doc/tutorial/tutorial.md]: Done

File: test/datalog_tests/simple.ast.expected (hunk 2 above)
Comment (ryzhyk) [reply test/datalog_tests/simple.ast.expected]: Now you're just nitpicking :)

**Pull request merged**